src/mc-reader-private.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MONGOCRYPT_READER_PRIVATE_H
+#define MONGOCRYPT_READER_PRIVATE_H
+
+#include "mongocrypt-buffer-private.h"
+#include <stdint.h>
+#include <stdlib.h>
+
+/**
+ * A non-owning forward-only cursor api to read a buffer.
+ *
+ * Tracks length of buffer and current position of buffer. parser_name is
+ * typically __FUNCTION__ to provide useful error messages automatically.
+ *
+ * All numbers are read as little endian.
+ *
+ * Functions return false on error and set mongocrypt_status_t to an error.
+ *
+ * Example:
+ *
+ *  const _mongocrypt_buffer_t *buf;
+ *  mongocrypt_status_t *status
+ *
+ *  mc_reader_t reader;
+ *  mc_reader_init_from_buffer(&reader, buf, __FUNCTION__);
+ *
+ *  _mongocrypt_buffer_t in;
+ *  if(!mc_reader_read_uuid_buffer(&reader, &in, status)) {
+ *    abort();
+ *  }
+ *
+ *  uint8_t value;
+ *  if(!mc_reader_read_u8(&reader, value, status )) {
+ *    abort();
+ *  }
+ */
+struct _mc_reader_t {
+   const uint8_t *ptr;
+   uint32_t pos;
+   uint32_t len;
```

Comment:
Consider making `pos` and `len` a `uint64_t`. Callers would not need to cast wider types.

Reply:
Done